library/Icingadb/Web/Controller.php

```diff
@@ -550,6 +550,86 @@ public function export(Query ...$queries)
             return true;
         }
 
+        if ($this->format === 'json' || $this->format === 'csv') {
+            $isJsonFormat = $this->format === 'json';
+            $response = $this->getResponse();
+
+            $fileName = $this->getRequest()->getControllerName();
+            if ($isJsonFormat) {
+                $response
+                    ->setHeader('Content-Type', 'application/json')
+                    ->setHeader('Cache-Control', 'no-store')
+                    ->setHeader(
+                        'Content-Disposition',
+                        'inline; filename=' . $fileName . '.json'
+                    )
+                    ->sendResponse();
+            } else {
+                $response
+                    ->setHeader('Content-Type', 'text/csv')
+                    ->setHeader('Cache-Control', 'no-store')
+                    ->setHeader(
+                        'Content-Disposition',
+                        'attachment; filename=' . $fileName . '.csv'
+                    )
+                    ->sendResponse();
+            }
+            // Disable output buffering to stream large amounts of data without a memory exhausted error.
+            ob_end_clean();
+            set_time_limit(0);
```

Comment:
Use the `Environment` class here instead. And don't let it run forever, use the default timeout of 300 seconds.

Reply:
300 is already default, but was not enough when streaming `json` for history (Maximum execution time of 300 seconds exceeded). 🤔

Reply:
Well. I'd say exporting the *entire* history isn't something that should be done anyway. Most users will still limit this to a reasonable size. It just shouldn't abort with a memory exhaustion error, a timeout error especially after five minutes is okay though.

Reply:
I see, but this means that the `json` stream stops after 5 minutes of execution and the trailing `]` or comma cannot be set if it is needed in between. 🤔

Reply:
It stops *nicely*? Isn't there an error message?

Reply:
Yes, there is. Ahh, I see. So, if it takes longer than 5 minutes, then it's OK to show this error and raw json.

Reply:
Well, the response code is 500, right? (I hope so and if so, it's fine)

Reply:
No it's not 500, headers are sent at this stage....

Reply:
Yeah, headers are sent, so there's no way to change the response code. Since we also have no special high level *protocol* here, other than the content type syntax, I see no other solution than running into the timeout and returning an error. It's no JSON or CSV after all, so anything processing the result should detect this and handle it accordingly.